nemo/collections/asr/models/ctc_bpe_models.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+import torch
+from omegaconf import MISSING, DictConfig, ListConfig, OmegaConf
+
+from nemo import logging
+from nemo.collections.asr.data.audio_to_text import AudioToBPEDataset
+from nemo.collections.asr.metrics.wer_bpe import WERBPE
+from nemo.collections.asr.models.ctc_models import EncDecCTCModel
+from nemo.collections.asr.parts.features import WaveformFeaturizer
+from nemo.collections.asr.parts.perturb import process_augmentations
+from nemo.collections.common import tokenizers
+from nemo.core.neural_types import *
+from nemo.utils.decorators import experimental
+
+__all__ = ['EncDecCTCModelBPE', 'JasperNetBPE', 'QuartzNetBPE']
+
+
+@experimental
+class EncDecCTCModelBPE(EncDecCTCModel):
+    """Encoder decoder CTC-based models with Byte Pair Encoding."""
+
+    def __init__(self, cfg: DictConfig, trainer=None):
+        if 'tokenizer' not in cfg:
+            raise ValueError("`cfg` must have `tokenizer` config to create a tokenizer !")
+
+        self.tokenizer_cfg = OmegaConf.to_container(cfg.tokenizer, resolve=True)  # type: dict
+        self.tokenizer_path = self.tokenizer_cfg.pop('path')  # Remove path and resolve based on tokenizer type
+
+        if os.path.exists(os.path.join(self.tokenizer_path, 'tokenizer.model')):
+            # This is a BPE Tokenizer
+            model_path = os.path.join(self.tokenizer_path, 'tokenizer.model')
+
+            if 'special_tokens' in self.tokenizer_cfg:
+                special_tokens = self.tokenizer_cfg['special_tokens']
+            else:
+                special_tokens = None
+
+            # Update special tokens
+            self.tokenizer = tokenizers.SentencePieceTokenizer(model_path=model_path, special_tokens=special_tokens)
```

Comment:
Perhaps we need advice from @ekmb @yzhang123 @VahidooX . Aren't HF's tokenizers currently the "best" (e.g. fast, flexible, used by other models)? Should we be choosing SentencePiece here?

Reply:
Had a discussion with Boris and Vitaly regarding this. We tried out the GPT2 BPE tokenizer built into HF, but it first pads the initial 256 encodings with the Unicode characters in range 0-255. This causes an artifical addition of tokens which are never trained (say on our preprocessed librispeech which is all lowercase 26 alphabet chars and space and apostrophe). We waste a lot of compute and memory in handling tokens that will never be trained at all on the LS corpus.

Vitaly used sentencepiece, and it seems other than a few (5) unique tokens, all other tokens are proper BPE tokens built form the corpus itself, so we plan to switch between WPE (also has correct behaviour) and BPE (sentencepiece). I've explicitly not added support for Hugging Face GPT2 (BPE) tokenizer due to this issue.